### ISSUE TITLE
Investigate add/pin/gc locking

### ISSUE THREAD

**Issue by lgierth** [ipfs/go-ipfs#3514]
While looking into #3470 and #3436, I found that the locks around adding/pinning/GC look really wonky and inconsistent.

This issue is a reminder that we need to:

- [ ] Look very closely at which locks are in play
- [ ] Document in detail how these locks work
- [ ] Write stress tests which make sure we don't end deadlock ourselves

**Comment by hsanjuan** [comment]
Yes, I've also noticed weird things like `pin ls` being blocked if a `pin add` is ongoing. I'm going to be looking more closely into this in the near future.
